aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -245,9 +245,6 @@ def validate_parameters(self):
         self.ctx.inputs.parameters = self.ctx.inputs.parameters.get_dict()
         self.ctx.inputs.settings = self.ctx.inputs.settings.get_dict() if 'settings' in self.ctx.inputs else {}
 
-        if 'parent_folder' in self.ctx.inputs:
-            self.ctx.restart_calc = self.ctx.inputs.parent_folder.creator
-
```

Comment:
The only place where `self.ctx.restart_calc` is checked is indeed in `prepare_process`. So it could potentially be fine to remove this, if indeed we do not have to explicitly set the `restart_mode` but as described in the other comment, in some cases we might, when the parameters are inconsistent. Alternatively, we could change the `restart_calc` to a boolean instead of a `CalcJobNode` since really we are merely using it as a flag and don't actually need any info from the creator calcjob.

Reply:
Hmm, do we even need the `restart_calc` then? Can't we just update the `self.ctx.inputs` whenever a calculation must restart? That way we can also differentiate between different modes of restarting more directly. For example:

1. For the `handle_out_of_walltime` process handler, we can just set the `restart_mode` to `'from_scratch'` in case the structure has changed, and make sure to `pop` any `startingpot` or `startingwfc` inputs.
2. If we want a full restart, we simply set `restart_mode` to `'restart'`.
3. If we want a restart from the charge density, as you would for example want in case the number of bands is changed by `sanity_check_insufficient_bands`, we set the `restart_mode` to `'from_scratch'` and `startingpot` to `'file'`.

Actually, maybe a clean solution would be to have a `restart_mode` setting in the context (perhaps defined as an enum?) that lists the different ways one can restart, and then in the `prepare_process` step this mode is used to set the correct inputs?

Reply:
Also see https://github.com/aiidateam/aiida-quantumespresso/issues/691, which is def relevant here.

Reply:
Yeah, the main point is to centralize the logic for dealing with restarts instead of having to deal with this each time in each handler explicitly. If an enum can cover all cases, that would be great, because that is easy to set. However, this might not be serializable and so may not be possible to be stored in the context.